File: 790.domino-and-tromino-tiling.py
#
# @lc app=leetcode id=790 lang=python3
#
# [790] Domino and Tromino Tiling
#

# @lc code=start
# ref: https://leetcode.com/problems/domino-and-tromino-tiling/description/comments/1566271
class Solution:
    def numTilings(self, n: int) -> int:
        # dp[1] = 1, dp[2] = 2, dp[3] = 1+2*(dp[2]-1)+(form by tromino)
        if n==1:
            return 1
        if n==2:
            return 2
        if n==3:
            return 5
        dp = [0]*(n+1)
        dp[1] = 1
        dp[2] = 2
        dp[3] = 5
        for i in range(4, n+1):
            dp[i] = 2*dp[i-1]+dp[i-3]
        return dp[n]%(10**9+7)
        
# @lc code=end

File: 790.domino-and-tromino-tiling.go
/*
 * @lc app=leetcode id=790 lang=golang
 *
 * [790] Domino and Tromino Tiling
 */

// @lc code=start
import "math"

// ans: https://leetcode.com/problems/domino-and-tromino-tiling/?envType=study-plan-v2&envId=leetcode-75
func numTilings(n int) int {
	// for n > 3, answer comes from a few sources
	// 1. ans[n-3] * 5
	// 2. ans[n-2] * 2
	// 3. ans[n-1] * 1
	// base cases,
	// n == 1: 1
	// n == 2: 2
	// n == 3: 5
	ans := []int{1, 2, 5}
	if n <= 3 {
		return ans[n-1]
	}
	for i := 3; i < n; i++ {
		best := (ans[i-3] + ans[i-1]*2) % int(math.Pow(10, 9)+7)
		ans = append(ans, best)
	}
	return ans[len(ans)-1]
}

// @lc code=end

